### DISCUSSION TITLE
How to pull messages via the DDS protocol without missing any

### DISCUSSION THREAD

**Discussion by tve** [opendcs/opendcs#814] [Q&A] [answered]
I'm writing a little service to pull messages from LRGS via DDS using the new python client (https://github.com/dcspy/dcspy) and I have a general question about the semantics around time.

I will be pulling messages for a group (at least what DADDS calls a group, is that a NETWORK_LIST in the search criteria or do I need to manage my own list of DCP addresses?) and I am wondering what I need to use for DRS_SINCE. Some options:
- use the timestamp of the most recent message received in the previous poll
- for each DCP keep track of the most recent timestamp, send the oldest one of these (issue with dead DCPs)
- use some fudge factor, e.g. now-1 hour and de-duplicate messages

Another way to ask is: if I have received a message with timestamp T for a specific DCP am I guaranteed that I will not, in the future, receive a message with timestamp <T?

A bit of context: this work is to support wildlife tracking receivers so there hopefully will be a large number of stations and the service I'm writing basically collects data from cdadata.wcda.noaa.gov, reformats it a bit, and pushes it to the Motus wildlife tracking system/database.

I apologize if my question is too far off-topic for OpenDCS.

**Comment by MikeNeilson** [comment]
Group/VS NETWORK_LIST - Probably. the network list for the LRGS is just a line separate list of stations IDs you want to receive. For GOES it's the NESDIS id.

For DRS_SINCE, The "session" is by connection, so if you connect and keep asking for data, it'll track where it is in the received data on the server side. However, if the connection fails, that session goes away, there's no other session tracking so it doesn't hurt to also track client side.

The majority of users use the "now - some amount" technique to deal with outage gaps. LRGS itself dedups messages, your client, or what ever uses the client, would need to do the same.

For the timestamp, no, that depends partly on the configuration of the LRGS and the search criteria. There's an option for "receive time" vs "data time". So it's possible for an LRGS to get a message late and still send it to clients. Don't assume the stream of messages is chronological.

Sounds like an interesting project.  You may be interested in this to have a non-network backup: https://github.com/opendcs/opendcs/wiki/Direct-GOES-data-using-cheap-tools (NOTE: I really need to finish writing that up but should get the idea across.)

**Reply by tve** [reply]
Thanks for the response! When you talk about session/connection, do you mean repeated queries or the "hang on the line" mode, i.e,. not specifying a DRS_UNTIL? I'm using "To request multiple DCP messages per request, the client sends a DDS message of type IdDcpBlock (‘n’)." and after getting the "ServerCode DUNTIL" I could issue the same command again. But I suspect you're referring to not sending a DRS_UNTIL, which is described as (from https://dcs1.noaa.gov/LRGS/DCP-Data-Service-14.pdf 6.1.1):

> If the search criteria contains no until time, this indicates that the client wishes to ‘hang
on the line’, retrieving data in real-time as it becomes available. When the server receives
an IdDcp request, AND no until time has been set, AND there are no new messages that
meet the client’s criteria, THEN the server MUST respond with an error message with
ServerCode = DMSGTIMEOUT (11). When the client receives this response, it
SHOULD pause briefly and then try the request again.

> The majority of users use the "now - some amount" technique to deal with outage gaps.

All makes sense to me and all stuff I can do. The part I don't know is how to size the "now - some amount" :-). Should I go back 24 hours? That could be a lot of data to throw away at each query. Should I go back 36 hours once every 12 hours to catch the laggards? Or do I need to go back several days once a day? Any advice based on experience would be appreciated!

> Group/VS NETWORK_LIST - Probably. the network list for the LRGS is just a line separate list of stations IDs you want to receive. For GOES it's the NESDIS id.

I just re-read section 5 on network lists in that DDS doc. So transient lists allow all the addresses to be uploaded once and then subsequent queries in one connection can just reference that. Is there a list or description of the _persistent_ lists somewhere? In my dream world those would correspond to what DADDS shows as groups, but I suspect that I'm dreaming...
The problem I'm trying to solve is to have to implement a mechanism for adding a new ID to the queries when someone brings a new transmitter online. If the transmitter could be part of some "group" and I query for that "group" then the new data would magically appear.

**Reply by tve** [reply]
> For the timestamp, no, that depends partly on the configuration of the LRGS and the search criteria. There's an option for "receive time" vs "data time". 

Thanks for pointing that out! If I can focus down on data transmitted through GOES DCS and then queried via the LRGS servers (`cdadata.wcda.noaa.gov` et. al.) I see three timestamps:
- DOMSAT header timestamp field
- DRS_SINCE=LGRS_SINCE in search criteria
- DAPS_SINCE in search criteria

I assume DRS_SINCE refers to the timestamp at which a LRGS server has received the message? And the DOMSAT header timestamp is when a message was received from GOES? And I'm learning that DAPS stands for DCS Automated Processing System which seems to be co-located with or include the LRGS servers and I'm wildly guessing that DAPS_SINCE effectively also refers to the reception time from GOES? Thus the timestamp in the DOMSAT header and in the DAPS_SINCE search criteria refer to the same timestamps? (phew)

Hmm, so if I understand correctly, if I run a query and get data and the query completes at wall clock time T, then I can say `LGRS_SINCE: T - 1 minute` in the next query and not miss anything. The response may contain data transmitted by GOES hours ago if there was an outage causing it to arrive late at the LRGS server, but it will forward it to me thanks to the search criteria. Yes, I need to do dedup and fit the late data into the data stream and all that fun stuff, but I will get it.

But if I round-robin across all the LRGS servers this strategy doesn't work. Then I should use DAPS_SINCE and go back N hours and dedup. So perhaps use DAPS_SINCE when querying a new LRGS server, then stick to that server and use LRGS_SINCE until "it's time" to move to the next server?

Does OpenDCS do all this automatically under the hood?

**Reply by MikeNeilson** [reply, marked as answer]
Okay, so here's where it might get confusing. I'm pretty sure you're statement about the times is correct. "DOMSAT" itself is defunct, but the concept in name lives on in the protocol and is when an LRGS/HRIT received the message. 

"DAPS" is the set of LRGSs that make up cdadata.wcda.noaa.gov.

Now as to how the streaming of messaging works.

The basic process is as follows:
1. Connect
2. Authenticate - currently optional
3. Send desired criteria
4. until error, or client decided to disconnect
  5. call "get next messages" 
  6. Process batch of messages

For the _SINCE fields, what that means is that the first batches of messages will be within that window. 

If you have an _UNTIL you'll get a message once all available messages that are in that window have been sent.
If you don't have an _UNTIL (or it's "realtime") then you'll get a message that "none are available" but you should ask again.

If you're familiar with dealing messages queues over HTTP it's basically "long-polling"

For the round-robin idea, that's interesting, but I'm not sure there's a good way to make it work with the way the criteria works. What the OpenDCS LRGS instance does is just take in all the data and checks for duplicates before it saves; also checks the quality flags and may overwrite if better. (This is more likely to happen if you're getting direct feed from the satellites though.)

An LRGS connected to an upstream LRGS, or the "RoutingScheduler" does this under the hood. This class, https://github.com/opendcs/opendcs/blob/main/java/opendcs/src/main/java/lrgs/ldds/LddsClient.java, is used by both to do that. It's not the easiest class to follow; a lot of the code was translated from C, and before good patterns were well established in Java. If I were doing this today that client would just return a `java.util.Stream` and in Python I'd use, I think yield, don't do python a lot but I think that's the equivalent concept.

One thing the java client does that it shouldn't, but makes sense in python, is using exceptions for flow control. That's all to just say if you copy the pattern of the client class above be judicious about how; it was written 20 years ago on a budget.

**Reply by MikeNeilson** [reply]
> Is there a list or description of the persistent lists somewhere? In my dream world those would correspond to what DADDS shows as groups, but I suspect that I'm dreaming...
The problem I'm trying to solve is to have to implement a mechanism for adding a new ID to the queries when someone brings a new transmitter online. If the transmitter could be part of some "group" and I query for that "group" then the new data would magically appear.

For this part, the "permanent" list is a bit of a side affect of the time the software was written. On the OpenDCS LRGS each user has a "home directory" where data can be saved. So you could upload a network list and then reference it by name. I'd recommend just ignoring it and always sending the network list on connection.

Future work on the LRGS may still support such operations, like improved user management, but we're not sure how that's going to pan out.

**Reply by tve** [reply]
Thanks so much for taking the time to answer in detail! I'll implement the "long-polling" in the python client. It's not clear to me yet how the SINCE criteria interacts with the notion of a per-session cursor. For now I'll assume the LRGS server does something like "if MSG_TS > SINCE and MSG_TS > CURSOR_TS" before forwarding a message. I'll find out as soon as I implement...

WRT persistent network lists I gather I should think of that as deprecated & avoid...

**Comment by agilmore2** [comment]
As far as round-robin goes, those connections will be completely
independent of eachother. The time windows and messages available should be
handled separately for every connection.

The round robin idea is implemented to handle complete failure of the
source, not temporary network blips etc.
